api/provision/csp-write-portfolio.ts

```diff
@@ -0,0 +1,142 @@
+  // we no longer need the mock implementations of CSP_<A|B|C|D> and can fully use the
+  // mock (or real) integration endpoints.
+  const cspResponse = await createCspResponse(stateInput);
+
+  // Throws a custom error identified by the state machine and the function retries 2
+  // times before failing continuing through the remaining states
+  if (cspResponse.code >= 500) {
+    const error = new Error(JSON.stringify(cspResponse));
+    Object.defineProperty(error, "name", {
+      value: "MockCspApiError",
+    });
+    throw error;
+  }
+
+  return cspResponse;
+}
+
+/**
+ * Make a request to an actual CSP implementation of the ATAT API
+ *
+ * @param request The input provisioning request
+ * @returns the response from the CSP
+ */
+async function makeARealRequest(request: ProvisionRequest): Promise<CspResponse> {
+  const baseUrl = (await getConfiguration(request.targetCsp.name))?.uri;
+  const url = `${baseUrl}/portfolios`;
+  if (!baseUrl || !baseUrl.startsWith("https://")) {
+    logger.error("Invalid CSP configuration", {
+      input: {
+        csp: request.targetCsp.name,
+      },
+      resolvedUrl: url,
+      configSecretPath: process.env.CSP_CONFIG_SECRET_NAME,
+    });
+    return {
+      code: 400,
+      content: {
+        details: "Invalid CSP provided",
+      },
+    };
+  }
+
+  const response = await axios.post(url, request.payload, {
+    headers: {
+      Authorization: `Bearer ${(await getToken()).access_token}`,
+      "User-Agent": "ATAT v0.2.0 client",
+    },
+    // Don't throw an error on non-2xx/3xx status code (let us handle it)
+    validateStatus() {
+      return true;
+    },
+  });
+
+  if (response.status !== 200 && response.status !== 202) {
+    logger.error("Request to CSP failed", {
+      csp: request.targetCsp.name,
+      request: {
+        csp: request.targetCsp.name,
+        url,
+      },
+      response: { statusCode: response.status, body: response.data },
+    });
+  }
+
+  return {
+    code: response.status,
+    content: response.data,
+  };
+}
+
+export async function createCspResponse(request: ProvisionRequest): Promise<CspResponse> {
+  let response: CspResponse;
+  switch (request.targetCsp.name) {
+    case "CSP_A":
+      response = {
+        code: 200,
+        content: {
+          some: "good content",
+        },
+      };
+      logger.info("Success response", { response: response as any });
+      return response;
+    case "CSP_B":
+      response = {
+        code: 400,
+        content: {
+          some: "bad content",
+        },
+      };
+      logger.error("Failed response", { response: response as any });
+      return response;
+    case "CSP_C":
+    case "CSP_D":
+      response = {
+        code: 500,
+        content: {
+          some: "internal error",
+        },
+      };
+      logger.error("Internal error response", { response: response as any });
+      return response;
+    default:
+      response = await makeARealRequest(request);
```

Comment:
Just making sure I am following - if we want to be able to test against our actual mock server, we can add it to our SM config and pass in `foo` as the CSP name, ensuring that we arrive at this clause, right?

Reply:
Exactly! The CSP name that I've configured so far is `CSP_Mock` but we can choose basically whatever we want. Once we iron out some things with the existing Mock server implementation (around our Auth issues), then I'll update this to just always `makeARealRequest`.

So for today:
 - `"CSP_A" | "CSP_B" | "CSP_C" | "CSP_D"` -> The same mock 200/400/500 responses as before
 - Any other value -> Lookup in config
   - If present (for now, just `"CSP_Mock"`), use the configured URI and send a request
   - Else, return a 400